ocaml/xapi/helpers.ml

```diff
@@ -1234,3 +1234,104 @@ let host_supports_hvm ~__context host =
+      * TODO: don't generate certificates in bash? *)
+    val gen_pem_if_not_exist : unit -> unit
+  end = struct
+    let xapi_stunnel_conf = "/etc/stunnel/xapi.conf"
+    let m = Mutex.create ()
+    let current_accept = ref None
+
+    let rewrite_xapi_ssl_config_file ~accept =
+      let idle_timeout = match Sys.getenv_opt "STUNNEL_IDLE_TIMEOUT" with
+                         | None -> "; no idle timeout"
+                         | Some x -> Printf.sprintf "TIMEOUTidle = %s" x
+      in
+      let conf_contents = String.concat "\n" [
+          "; autogenerated by helpers.ml"
+        ; "fips = no"
+        ; "pid = /var/run/xapissl.pid"
+        ; "socket = r:TCP_NODELAY=1"
+        ; "socket = a:TCP_NODELAY=1"
+        ; "socket = l:TCP_NODELAY=1"
+        ; "socket = r:SO_KEEPALIVE=1"
+        ; "socket = a:SO_KEEPALIVE=1"
+        ; idle_timeout
+        ; "debug = authpriv.5"
+        ; ""
+        ; "[xapi]"
+        ; Printf.sprintf "accept = %s" accept
+        ; "connect = 80"
+        ; Printf.sprintf "cert = %s" cert
+        ; Printf.sprintf "ciphers = %s" Xcp_const.good_ciphersuites
+        ; "curve = secp384r1"
+        ; "TIMEOUTclose = 1"
+        ; "options = CIPHER_SERVER_PREFERENCE"
+        ; "sslVersion = TLSv1.2"
+        ]
+      in
+      if Sys.file_exists xapi_stunnel_conf then Sys.remove xapi_stunnel_conf;
+      Stdext.Unixext.write_string_to_file xapi_stunnel_conf conf_contents
+
+    let update_config ~accept =
+      Mutex.execute m (fun () ->
+        match !current_accept with
+        | Some current_accept when current_accept = accept -> ()
+        | None | Some _ -> begin
+          current_accept := Some accept;
+          rewrite_xapi_ssl_config_file ~accept
+        end
+      )
+
+    let gen_pem_if_not_exist () =
+      Mutex.execute m (fun () ->
+        if not (Sys.file_exists cert) then begin
+          let hostname = call_script "/usr/bin/hostname" ["-f"] |> Astring.String.trim in
+          let cn =
+            if Astring.String.is_prefix ~affix:"localhost" hostname then
+              "mgmt_ip"
+            else if Astring.String.is_infix ~affix:"." hostname then
+              hostname
+            else
+              "mgmt_ip"
+          in
+          call_script "/opt/xensource/libexec/generate_ssl_cert" [cert; cn] |> ignore
+        end
+      )
+
+  end
+
+  exception Xapi_stunnel_conf_doesnt_exist
+
+  let systemctl cmd = call_script "/usr/bin/systemctl" [cmd; "stunnel@xapi"]
+  let systemctl_ cmd = systemctl cmd |> ignore
+
+  let is_enabled () =
+    systemctl "is-enabled" |>
+    Astring.String.trim |>
+    function
+    | "enabled"  -> true
+    | "disabled" -> false
+    | _          -> D.error "Stunnel.is_enabled: unable to discern if stunnel@xapi is enabled. assuming that it is disabled"; false
+
+  let restart ~accept =
+    File.update_config ~accept;
+    File.gen_pem_if_not_exist ();
+    if not @@ is_enabled () then systemctl_ "enable";
+    systemctl_ "restart"
```

Comment:
I think we might have to modify the default stunnel service file via systemd drop-ins, the old startup script also had this:
```
    # Increase the fd limit to 2048, so stunnel could accept 1000 client connections
    ulimit -n 2048
```

That one now has to be set using systemd directives.